### ISSUE TITLE
Issues and possible improvements to transparency rendering

### ISSUE THREAD

**Issue by mvaligursky** [playcanvas/engine#5943]
The default framebuffer WebGL uses for rendering is RGB(A)8 buffer. The values written to it are expected to be in the display color space, which typically means sRGB. As this framebuffer is created by the canvas, we have no control over it, and it is not created with sRGB (EXT_sRGB) flag. Meaning the shader writing to it needs to manually encode values from linear to sRGB space.

### LDR rendering
- when rendering directly to the default framebuffer, the engine correctly encodes values to sRGB space at the end of shaders. This causes alpha blending to not be correct, as the blending takes place in sRGB space, instead of correctly in linear space, with the result getting encoded to sRGB. This is how PlayCanvas currently operates if rendering directly to the default framebuffer.
- proposal to support linear workflow: https://www.w3.org/Graphics/Color/Workshop/slides/Russell.pdf (it does not seem to be progressing, see https://groups.google.com/a/chromium.org/g/blink-dev/c/tA7Av4fLQFY/m/oUw_k949AAAJ). 
- a workaround is rendering to a created framebuffer using sRGB, which would alpha blend correctly, and blit the framebuffer to the default framebuffer at the end of the frame. This is expensive, and so not typically used.
- we also implement dithered transparency. As this does not depend on alpha blending, it's not affected by blending in gamma space and works correctly. The only problem is that when enabled, this looks visually different to alpha blending, as that is incorrect.

### HDR rendering
- When we render in HDR mode, we use float format of the internal render target in linear space. This makes alpha blending to work in linear space, which is correct.
- We have problem with transparency dithering here if tone mapping is used - it does not match the correct alpha blending. The problem is that with dithering, the discarded and visible pixels are tone mapped, but as tone-mapping is not linear, it does not match the alpha blending,  as non-blended values are tonemapped.
- Transparency dithering works correctly if tone mapping is used, as long as TAA is used as well - as TAA resolves transparency by blending with history buffer, and this blending takes place before the tone mapping.

**References:**
https://github.com/KhronosGroup/WebGL/issues/2474
https://github.com/mrdoob/three.js/issues/23019
https://stackoverflow.com/questions/51032480/how-do-you-implement-a-gamma-correct-workflow-in-webgl2
https://groups.google.com/g/angleproject/c/5ITMg4_m8Ug
https://github.com/playcanvas/engine/issues/5287

**Comment by MAG-AdrianMeredith** [comment]
its not necessarily srgb though right? https://ccameron-chromium.github.io/webgl-examples/p3.html. but I guess this would actually help here as we're letting the hardware do it

**Comment by mvaligursky** [comment]
True, but until we implement https://github.com/playcanvas/engine/issues/5287, it's sRGB. Handling P3 is part of that issue.

**Comment by mvaligursky** [comment]
Related Chromium issue I created exploring the sRGB in WebXR: https://partnerissuetracker.corp.google.com/issues/352345526?pli=1

**Comment by mvaligursky** [comment]
conclusion regarding WebXR, from the issue above:

_The mechanism for specifying a texture format for WebXR is the_ [Layers API](https://urldefense.proofpoint.com/v2/url?u=https-3A__immersive-2Dweb.github.io_layers_-23xrprojectionlayerinittype&d=DwMFaQ&c=ncDTmphkJTvjIDPh0hpF_4vCHvabgGkICC2epckfdiw&r=fcHApDUmjByu7sKe8PjfyjZ4Bea75Kb0SPd-OTdqx1U&m=qtKDJRJeTuLVdv3E6OTUIEED8Smqve3DGZ2T9BJ7jqbLtifytBjc30sVV3xVNgp3&s=bDP8iEfDHvKCU3KLPuJcLzbQ36Ym9fWBbXTL60J3GEI&e=), _which is not yet supported by Chrome but does have support in the Meta Quest browser. WebXR is not compatible with buffers allocated by drawingBufferStorage._

**Comment by mvaligursky** [comment]
https://github.com/playcanvas/engine/pull/6838 implements support for sRGB framebuffer on WebGPU, but also support for sRGB render target on all platforms (just not the default framebuffer on WebGL), allowing linear rendering.

**Comment by mvaligursky** [comment]
Update on this. 
- the linear blending can be used on both WebGL and WebGPU if the color buffer of a render target uses some of the sRGB formats, for example SRGBA8
- sRGBA8 format can be used as a canvas framebuffer format on WebGPU as well, using `displayFormat` member of `createGraphicsDevice` options
- on WebGL the support is not implemented currently and the options are more limited:
    - drawingBufferStorage can be used to set it, but this is only supported on Chrome 122+, other browsers do not implement it yet
    - Layers API can be used to set this for WebXR, but Chrome does not implement this (see more above)

**Comment by mvaligursky** [comment]
relevant topic: https://www.fractolog.com/2024/07/color-space-correctness-in-alpha-blending/

**Comment by animech-sijo** [comment]
Has there been any progress made on this issue?

We're running into problems when using HDR post processing effects (such as bloom). When using the built in `pc.CameraFrame`,  transparent materials render differently than the editor (even when no post processing effects are used). 

To test this, we set up similar scenes in different rendering engines. We made the background solid white, used linear tonemapping, created a solid black material (black diffuse, no specular reflections) and added a linear 8 bit gradient texture in the opacity slot. The opacity texture had its sRGB flag unchecked. We applied the material to a plane, and added a camera looking straight down onto the plane. Our Playcanvas test scene can be found [here](https://playcanvas.com/editor/scene/2410142). Here are some results:

| Engine      | Gradient      |
| ------------- | ------------- |
| PlayCanvas (default/editor) | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/8a35ea32-ede1-4628-b873-b157185bc0ef" /> |
| PlayCanvas (pc.CameraFrame) | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/67a596b7-aa93-4a18-8b51-3bae898d3e2a" /> |
| Blender | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/52258df8-32b2-4c21-9dc5-13082e512eaf" /> |
| Unity | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/6b2ec558-b088-45cc-9c21-6b3b4d62a459" /> |

When compared to other rendering engines, we see that Playcanvas (by default, without post processing) blends transparent materials in the wrong color space. When rendering in HDR using `pc.CameraFrame`, we get the correct blending. However, we now have a discrepancy with the editor since it renders without post processing.

Artists primarily use the editor's viewport when working, and need it to match the runtime results. And this shows that the editor (which uses the default engine) blends transparency wrong.

**Comment by Maksims** [comment]
This problem is extra exaugurated when application offers quality presets.
We have a project where in Low quality we render directly to backbuffer, while in high quality we enable CameraFrame and render into its framebuffer, in HDR.
Switching between them noticeably affects alpha blended colors.

**Comment by mvaligursky** [comment]
Just having a quick look, it seems drawingBufferStorage is supported by Chromium browsers already, but not other browsers.
A solution could be to expose `Linear Blending` setting somewhere in Editor Render Settings. When enabled, both the Editor and the runtime would:
- use drawingBufferStorage when available
- do a full pass blit when not available (this is a high cost, especially on mobile).

WebGPU natively supports this, so no issues there.

**Comment by Maksims** [comment]
Could there be a user control, if they are fine for "paying" the blit cost - one way. But also provide an option to enable such only when it is available, and app might be ok with visual difference, and does not want to have expensive blit. Also an ability to force blit path for testing. This is pretty advanced users feature.
Hopefully one day it will all rendered to HDR natively.

**Comment by animech-sijo** [comment]
Yes a user option would be preferable. I'm also thinking how older projects should be handled. All their textures are authored with the incorrect gamma. So it feels like that this feature should be opt-in for older projects.
